File: Delivery System 2/delivery_system.cpp
#include "delivery_system.h"
#include <iostream>
#include <algorithm>

// Èíèöèàëèçàöèÿ ñòàòè÷åñêèõ ïåðåìåííûõ
int DeliverySystem::s_globalOrderCount = 0;
const std::string DeliverySystem::s_systemVersion = "2.1.0";

// Èíèöèàëèçàöèÿ ñòàòè÷åñêîé ïåðåìåííîé
int DeliverySystem::m_orderCounter = 0;

// Êîíñòðóêòîð
DeliverySystem::DeliverySystem() {}

// Äåñòðóêòîð
DeliverySystem::~DeliverySystem() {
}

// Âûâîä èíôîðìàöèè î ñèñòåìå ñ èñïîëüçîâàíèåì this
void DeliverySystem::printSystemInfo() const {
	std::cout << "=== Èíôîðìàöèÿ î ñèñòåìå äîñòàâêè ===" << std::endl;
	std::cout << "Âåðñèÿ ñèñòåìû: " << this->getSystemVersion() << std::endl;
	std::cout << "Âñåãî çàêàçîâ â ñèñòåìå: " << this->m_allOrders.size() << std::endl;
	std::cout << "Âñåãî êëèåíòîâ: " << this->m_allClients.size() << std::endl;
	std::cout << "Âñåãî êóðüåðîâ: " << this->m_allCouriers.size() << std::endl;
	std::cout << "Ãëîáàëüíûé ñ÷åò÷èê çàêàçîâ: " << this->getGlobalOrderCount() << std::endl;
}

// Ñîçäàíèå íîâîãî çàêàçà ñ îáðàáîòêîé èñêëþ÷åíèé
std::shared_ptr<Order> DeliverySystem::createOrder(std::shared_ptr<Client> sender, std::shared_ptr<Client> receiver, const Address& from, const Address& to, const Parcel& parcel, std::shared_ptr<Tariff> tariff)
{
	try {
		// Âàëèäàöèÿ âõîäíûõ äàííûõ
		validateOrderData(from, to, parcel);

		if (!tariff) {
			throw std::invalid_argument("Òàðèô íå ìîæåò áûòü íóëåâûì");
		}
	std::string trackingNumber = "TRK" + std::to_string(++m_orderCounter);
	auto newOrder = std::make_shared<Order>(trackingNumber, from, to, parcel, tariff.get());

	m_allOrders.push_back(newOrder);
	if (sender) sender->addToHistory(newOrder);

	incrementGlobalOrderCount();

	return newOrder;
}
	catch (const InvalidAddressException& e) {
		std::cerr << "Îøèáêà àäðåñà ïðè ñîçäàíèè çàêàçà: " << e.what() << std::endl;
		throw OrderCreationException("Íå óäàëîñü ñîçäàòü çàêàç èç-çà íåâåðíîãî àäðåñà", e);
	}
	catch (const InvalidParcelException& e) {
		std::cerr << "Îøèáêà ïîñûëêè ïðè ñîçäàíèè çàêàçà: " << e.what() << std::endl;
		throw OrderCreationException("Íå óäàëîñü ñîçäàòü çàêàç èç-çà íåâåðíûõ äàííûõ ïîñûëêè", e);
	}
	catch (const std::invalid_argument& e) {
		std::cerr << "Íåâåðíûé àðãóìåíò ïðè ñîçäàíèè çàêàçà: " << e.what() << std::endl;
		throw OrderCreationException("Íåâåðíûå ïàðàìåòðû çàêàçà", e);
	}
	catch (const std::exception& e) {
		std::cerr << "Íåèçâåñòíàÿ îøèáêà ïðè ñîçäàíèè çàêàçà: " << e.what() << std::endl;
		throw OrderCreationException("Íåèçâåñòíàÿ îøèáêà ïðè ñîçäàíèè çàêàçà", e);
	}
}

// Ïîèñê êëèåíòà ïî ID ñ èñêëþ÷åíèåì
std::shared_ptr<Client> DeliverySystem::findClientById(int id) const {
    for (const auto& client : m_allClients) {
        if (client->getId() == id) {
            return client;
        }
    }
    throw ClientNotFoundException("Êëèåíò ñ ID " + std::to_string(id) + " íå íàéäåí");
}

// Ïîèñê äîñòóïíîãî êóðüåðà ñ èñêëþ÷åíèåì
std::shared_ptr<Courier> DeliverySystem::findAvailableCourier() const {
    for (const auto& courier : m_allCouriers) {
        if (courier->getIsAvailable()) {
            return courier;
        }
    }
    throw NoAvailableCourierException("Íåò äîñòóïíûõ êóðüåðîâ â ñèñòåìå");
}

// Âàëèäàöèÿ äàííûõ çàêàçà
void DeliverySystem::validateOrderData(const Address& from, const Address& to, const Parcel& parcel) const {
    // Ïðîâåðêà àäðåñà îòïðàâèòåëÿ
    if (from.getStreet().empty() || from.getCity().empty()) {
        throw InvalidAddressException("Àäðåñ îòïðàâèòåëÿ íå ìîæåò áûòü ïóñòûì");
    }

    // Ïðîâåðêà àäðåñà ïîëó÷àòåëÿ
    if (to.getStreet().empty() || to.getCity().empty()) {
        throw InvalidAddressException("Àäðåñ ïîëó÷àòåëÿ íå ìîæåò áûòü ïóñòûì");
    }

    // Ïðîâåðêà ÷òî àäðåñà ðàçíûå
    if (from.getStreet() == to.getStreet() && from.getCity() == to.getCity()) {
        throw InvalidAddressException("Àäðåñà îòïðàâèòåëÿ è ïîëó÷àòåëÿ íå ìîãóò ñîâïàäàòü");
    }

    // Ïðîâåðêà ïîñûëêè
    if (parcel.getWeight() <= 0) {
        throw InvalidParcelException("Âåñ ïîñûëêè äîëæåí áûòü ïîëîæèòåëüíûì");
    }

    if (parcel.getWeight() > 100.0) {
        throw InvalidParcelException("Âåñ ïîñûëêè íå ìîæåò ïðåâûøàòü 100 êã");
    }

    if (parcel.calculateVolume() <= 0) {
        throw InvalidParcelException("Îáúåì ïîñûëêè äîëæåí áûòü ïîëîæèòåëüíûì");
    }

    if (parcel.calculateVolume() > 1000000.0) { // 1 ì?
        throw InvalidParcelException("Îáúåì ïîñûëêè íå ìîæåò ïðåâûøàòü 1 ì?");
    }

    if (parcel.getEstimatedValue() < 0) {
        throw InvalidParcelException("Îöåíî÷íàÿ ñòîèìîñòü íå ìîæåò áûòü îòðèöàòåëüíîé");
    }
}

// Ïîèñê çàêàçîâ ïî ñòàòóñó
std::vector<std::shared_ptr<Order>> DeliverySystem::findOrdersByStatus(OrderStatus status) const {
	std::vector<std::shared_ptr<Order>> result;
	for (auto order : m_allOrders) {
		if (order->getStatus() == status) {
			result.push_back(order);
		}
	}
	return result;
}

// Äîáàâëåíèå êëèåíòà â ñèñòåìó
void DeliverySystem::addClient(std::shared_ptr<Client> client) {
	m_allClients.push_back(client);
}

// Äîáàâëåíèå êóðüåðà â ñèñòåìó
void DeliverySystem::addCourier(std::shared_ptr<Courier> courier) {
	m_allCouriers.push_back(courier);
}

// Äîáàâëåíèå òàðèôà â ñèñòåìó
void DeliverySystem::addTariff(std::shared_ptr<Tariff> tariff) {
	m_availableTariffs.push_back(tariff);
}

File: Delivery System 2 Java/src/DeliverySystem.java
package src;

import java.util.ArrayList;
import java.util.List;

public class DeliverySystem {
    private List<Order> allOrders;
    private List<Courier> allCouriers;
    private List<Client> allClients;
    private List<Warehouse> allWarehouses;
    private List<Tariff> availableTariffs;
    private static int orderCounter = 0;

    // Статические поля
    private static int globalOrderCount = 0;
    private static final String SYSTEM_VERSION = "1.0.0";

    // Конструктор
    public DeliverySystem() {
        this.allOrders = new ArrayList<>();
        this.allCouriers = new ArrayList<>();
        this.allClients = new ArrayList<>();
        this.allWarehouses = new ArrayList<>();
        this.availableTariffs = new ArrayList<>();
    }

        // Статические методы
    public static int getGlobalOrderCount() {
        return globalOrderCount;
    }
    
    public static String getSystemVersion() {
        return SYSTEM_VERSION;
    }
    
    public static void incrementGlobalOrderCount() {
        globalOrderCount++;
    }
    
    public static void resetGlobalStatistics() {
        globalOrderCount = 0;
    }

    // Метод с использованием this
    public DeliverySystem getSystemInstance() {
        return this;
    }
    
    public void printSystemInfo() {
        System.out.println("=== Информация о системе доставки (использование this) ===");
        System.out.println("Версия системы: " + DeliverySystem.getSystemVersion());
        System.out.println("Всего заказов в этой системе: " + this.allOrders.size());
        System.out.println("Всего клиентов: " + this.allClients.size());
        System.out.println("Всего курьеров: " + this.allCouriers.size());
        System.out.println("Глобальный счетчик заказов: " + DeliverySystem.getGlobalOrderCount());
    }

    // Основные методы
    public Order createOrder(Client sender, Client receiver, Address from, 
                            Address to, Parcel parcel, Tariff tariff) {
        String trackingNumber = "TRK" + (++orderCounter);
        Order newOrder = new Order(trackingNumber, from, to, parcel, tariff);
        
        allOrders.add(newOrder);
        if (sender != null) {
            sender.addToHistory(newOrder);
        }
        
        // Увеличиваем глобальный счетчик
        incrementGlobalOrderCount();

        return newOrder;
    }

        // Методы с обработкой исключений
    public Order createOrderWithValidation(Client sender, Client receiver, Address from, 
                                         Address to, Parcel parcel, Tariff tariff) 
            throws InvalidAddressException, InvalidParcelException {
        
        // Валидация данных
        validateOrderData(from, to, parcel);
        
        if (tariff == null) {
            throw new IllegalArgumentException("Tariff cannot be null");
        }

        String trackingNumber = "TRK" + (++orderCounter);
        Order newOrder = new Order(trackingNumber, from, to, parcel, tariff);
        
        allOrders.add(newOrder);
        if (sender != null) {
            sender.addToHistory(newOrder);
        }
        
        incrementGlobalOrderCount();
        return newOrder;
    }

    // Валидация данных с выбрасыванием исключений
    private void validateOrderData(Address from, Address to, Parcel parcel) 
            throws InvalidAddressException, InvalidParcelException {
        
        // Проверка адресов
        if (from.getStreet() == null || from.getStreet().trim().isEmpty()) {
            throw new InvalidAddressException("Улица отправителя не может быть пустой");
        }
        
        if (to.getStreet() == null || to.getStreet().trim().isEmpty()) {
            throw new InvalidAddressException("Улица получателя не может быть пустой");
        }
        
        if (from.getCity() == null || from.getCity().trim().isEmpty()) {
            throw new InvalidAddressException("Город отправителя не может быть пустым");
        }
        
        if (to.getCity() == null || to.getCity().trim().isEmpty()) {
            throw new InvalidAddressException("Город получателя не может быть пустым");
        }
        
        // Проверка что адреса разные
        if (from.getStreet().equals(to.getStreet()) && from.getCity().equals(to.getCity())) {
            throw new InvalidAddressException("Адреса отправителя и получателя не могут совпадать");
        }
        
        // Проверка посылки
        if (parcel.getWeight() <= 0) {
            throw new InvalidParcelException("Вес посылки должен быть положительным");
        }
        
        if (parcel.getWeight() > 100.0) {
            throw new InvalidParcelException("Вес посылки не может превышать 100 кг");
        }
        
        if (parcel.calculateVolume() <= 0) {
            throw new InvalidParcelException("Объем посылки должен быть положительным");
        }
        
        if (parcel.calculateVolume() > 1000000.0) {
            throw new InvalidParcelException("Объем посылки не может превышать 1 м^3");
        }
        
        if (parcel.getEstimatedValue() < 0) {
            throw new InvalidParcelException("Оценочная стоимость посылки не может быть отрицательной");
        }
    }

    // Поиск с исключением
    public Client findClientById(int id) throws ClientNotFoundException {
        for (Client client : allClients) {
            if (client.getId() == id) {
                return client;
            }
        }
        throw new ClientNotFoundException("Клиент с ID " + id + " не найден");
    }

    public Courier findAvailableCourier() throws NoAvailableCourierException {
        for (Courier courier : allCouriers) {
            if (courier.getIsAvailable()) {
                return courier;
            }
        }
        throw new NoAvailableCourierException("Нет доступных курьеров в системе");
    }

    // Метод с try-catch блоком
    public void processOrderSafely(Client sender, Client receiver, Address from, 
                                  Address to, Parcel parcel, Tariff tariff) {
        try {
            Order order = createOrderWithValidation(sender, receiver, from, to, parcel, tariff);
            System.out.println("Заказ успешно создан: " + order.getTrackingNumber());
            
        } catch (InvalidAddressException e) {
            System.err.println("Ошибка валидации адреса: " + e.getMessage());
        } catch (InvalidParcelException e) {
            System.err.println("Ошибка валидации посылки: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("Неожиданная ошибка: " + e.getMessage());
        }
    }

    public List<Order> findOrdersByStatus(OrderStatus status) {
        List<Order> result = new ArrayList<>();
        for (Order order : allOrders) {
            if (order.getStatus() == status) {
                result.add(order);
            }
        }
        return result;
    }

    // Методы добавления сущностей
    public void addClient(Client client) {
        allClients.add(client);
    }

    public void addCourier(Courier courier) {
        allCouriers.add(courier);
    }

    public void addTariff(Tariff tariff) {
        availableTariffs.add(tariff);
    }

    // Геттеры
    public List<Order> getAllOrders() { return allOrders; }
    public List<Courier> getAllCouriers() { return allCouriers; }
    public List<Client> getAllClients() { return allClients; }
    public List<Tariff> getAvailableTariffs() { return availableTariffs; }
}